safeeyes/model.py

```diff
@@ -115,82 +125,86 @@ def is_long_break(self):
-        if self.__current_break is not None:
-            # Reset the time of long breaks
-            if self.__current_break.type == BreakType.LONG_BREAK:
-                self.__current_break.time = self.__long_break_time
+            break_obj = self.__next_short()
+
+        if self.__shorts_taken == len(self.__short_queue):
+            self.context['new_cycle'] = True
+            self.__shorts_taken = 0
+        else:
+            self.context['new_cycle'] = False
 
         self.__current_break = break_obj
         self.context['session']['break'] = self.__current_break.name
 
         return break_obj
 
     def reset(self):
-        self.__short_pointer = self.__short_header
-        self.__long_pointer = self.__long_header
-        self.__first_break = None
-        self.__current_break = None
-
-        # Reset all break time
-        short_pointer = self.__short_pointer
-        long_pointer = self.__long_pointer
-        short_pointer.time = self.__short_break_time
-        long_pointer.time = self.__long_break_time
-        short_pointer = short_pointer.next
-        long_pointer = long_pointer.next
-
-        while short_pointer != self.__short_header:
-            short_pointer.time = self.__short_break_time
-            short_pointer = short_pointer.next
+        for break_object in self.__short_queue:
+            break_object.time = self.__short_break_time
+        
+        for break_object in self.__long_queue:
+            break_object.time = self.__long_break_time
 
-        while long_pointer != self.__long_header:
-            long_pointer.time = self.__long_break_time
-            long_pointer = long_pointer.next
+    def is_empty(self):
+        return self.__short_queue is None and self.__long_queue is None
+
+    def __next_short(self, ignore_time=False):
+        longs  = self.__long_queue
+        shorts = self.__short_queue
+        break_obj = shorts[self.__current_short]
+        self.context['break_type'] = 'short'
+        # Reduce the break time from the next long break (default)
+        if not ignore_time:
+            longs[self.__current_long].time -= shorts[self.__current_short].time
+
+        # Update the index to next
+        if self.__is_random_order:
+            self.__current_short = random.randint(0, len(shorts) - 1)
+        else:
+            self.__current_short = (self.__current_short + 1) % len(shorts)
+        self.__shorts_taken += 1
+        return break_obj
 
+    def __next_long(self):
+        longs  = self.__long_queue
+        shorts = self.__short_queue
+        longs[self.__current_long].time = self.__long_break_time
+        break_obj = longs[self.__current_long]
+        self.context['break_type'] = 'long'
 
-    def is_empty(self):
-        return self.__short_pointer is None and self.__long_pointer is None
+        # Update the index to next
+        if self.__is_random_order:
+            self.__current_long = random.randint(0, len(longs) - 1)
```

Comment:
What do you think about choosing a random break from the remaining breaks to give all of them a chance. According to the current logic, there is a chance to get the same break more than once. In worst case, we may get the same break again and again (though the probability is much less)

I think it we can track the taken breaks and consider the remaining breaks only for the next random selection, the order will be random but all breaks will be guaranteed to be chosen.

WDYT?

Reply:
Same for short break random selection

Reply:
Will do